lib/Basics/DebugRaceController.cpp

```diff
@@ -24,47 +24,59 @@
 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE
 
 #include "DebugRaceController.h"
+
 #include "ApplicationFeatures/ApplicationServer.h"
+#include "Basics/ScopeGuard.h"
 
 using namespace arangodb;
 using namespace arangodb::basics;
 
 static DebugRaceController Instance{};
 
-DebugRaceController::DebugRaceController() {}
-
 DebugRaceController& DebugRaceController::sharedInstance() { return Instance; }
 
 void DebugRaceController::reset() {
-  std::unique_lock<std::mutex> guard(_mutex);
+  {
+    auto guard = std::scoped_lock<std::mutex>(_mutex);
+    _data.clear();
+  }
   _condVariable.notify_all();
-  _data.clear();
-  _didTrigger = false;
 }
 
-std::vector<std::any> DebugRaceController::data() const {
-  std::unique_lock<std::mutex> guard(_mutex);
-  return _data;
+bool DebugRaceController::didTrigger(
+    std::unique_lock<std::mutex> const& guard,
+    std::size_t numberOfThreadsToWaitFor) const {
+  TRI_ASSERT(guard.owns_lock());
+  TRI_ASSERT(_data.size() <= numberOfThreadsToWaitFor);
+  return numberOfThreadsToWaitFor == _data.size();
 }
 
 auto DebugRaceController::waitForOthers(
     size_t numberOfThreadsToWaitFor, std::any myData,
-    arangodb::application_features::ApplicationServer const& server) -> bool {
-  std::unique_lock<std::mutex> guard(_mutex);
-  if (!_didTrigger) {
+    application_features::ApplicationServer const& server)
+    -> std::optional<std::vector<std::any>> {
+  auto notifyGuard =
+      scopeGuard([this]() noexcept { _condVariable.notify_all(); });
+  {
+    std::unique_lock<std::mutex> guard(_mutex);
+
+    if (didTrigger(guard, numberOfThreadsToWaitFor)) {
+      return std::nullopt;
```

Comment:
Why are we returning `std::nullopt` here in case didTrigger() returns true, but return `_data` below if didTrigger() returns true there? This may be correct, but I don't understand it.

Reply:
The goal is to get exactly `numberOfThreadsToWaitFor` to get into the code below, add their data, and then continue. Any thread above that number will run into the early return here.

The check below is for a different reason: If the wait finishes, usually `didTrigger` is true, i.e. `_data.size() == numberOfThreadsToWaitFor`, and we return the data. But if the wait is aborted early, e.g. due to a reset (which clears `_data`), then we return `nullopt` instead.